step-plans/step-plans-base-artefacts/src/main/java/step/artefacts/handlers/functions/TokenForecastingContext.java

```diff
@@ -9,64 +11,104 @@
+        String poolName = bestMatchingPool.name;
+
+        // Delegate the creation of the provisioning parameters map to the registered parameter types
+        HashMap<String, String> provisioningParameters = new HashMap<>();
+        supportedParameters.forEach(p -> p.tokenSelectionCriteriaToAgentPoolProvisioningParameters.accept(criteria, provisioningParameters));
+
+        Key key = new Key(poolName, provisioningParameters, criteria);
+        requireToken(key, count);
+        return key;
+    }
+
+    protected void requireToken(Key key, int count) {
+        poolResourceReservations.computeIfAbsent(key, k -> new PoolReservationTracker()).reserve(count);
     }
 
-    protected void requireToken(String pool, int count) {
-        poolResourceReservations.computeIfAbsent(pool, k -> new PoolReservationTracker()).reserve(count);
+    protected static class Key {
+        String poolTemplateName;
+        Map<String, String> provisioningParameters;
+        Map<String, Interest> criteria;
+
+        public Key(String poolTemplateName, Map<String, String> provisioningParameters, Map<String, Interest> criteria) {
+            this.poolTemplateName = poolTemplateName;
+            this.provisioningParameters = provisioningParameters;
+            this.criteria = criteria;
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            if (this == o) return true;
+            if (o == null || getClass() != o.getClass()) return false;
+            Key key = (Key) o;
+            return Objects.equals(poolTemplateName, key.poolTemplateName) && Objects.equals(provisioningParameters, key.provisioningParameters);
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(poolTemplateName, provisioningParameters);
+        }
     }
 
-    protected void releaseRequiredToken(String pool, int count) {
-        poolResourceReservations.computeIfAbsent(pool, k -> new PoolReservationTracker()).release(count);
+    protected void releaseRequiredToken(Key key, int count) {
+        poolResourceReservations.computeIfAbsent(key, k -> new PoolReservationTracker()).release(count);
     }
 
-    private String getBestMatchingPool(Map<String, Interest> criteria) {
-        SimpleAffinityEvaluator<Identity, Identity> affinityEvaluator = new SimpleAffinityEvaluator<>();
-        return pools.entrySet().stream()
-                .map(entry -> new Object[]{affinityEvaluator.getAffinityScore(new TokenPretender(Map.of(), criteria), new TokenPretender(entry.getValue(), Map.of())), entry.getKey()})
-                .filter(o -> ((int) o[0]) >= 0).sorted(Comparator.comparingInt(o -> (int) o[0])).map(o -> (String) o[1])
-                .findFirst().orElse(null);
+    private Optional<AgentPoolSpec> getBestMatchingPool(Map<String, Interest> criteria) {
+        PreProvisioningTokenAffinityEvaluator affinityEvaluator = new PreProvisioningTokenAffinityEvaluator();
+        return availableAgentPools.stream()
```

Comment:
I got stuck a while to review this piece of code. At least some comment should be added, a refactoring would be nice if possible.